text/0000-threadsafe-callback.md

```diff
@@ -0,0 +1,123 @@
+unsafe impl Sync for ThreadSafeCallbackInner {}
+
+impl Drop for ThreadSafeCallbackInner {
+    fn drop(&mut self) {
+        // free resources
+    }
+}
+
+#[derive(Clone)]
+pub struct ThreadSafeCallback(Arc<ThreadSafeCallbackInner>);
+
+impl ThreadSafeCallback {
+    pub fn new<T: Value>(this: Handle<T>, function: Handle<JsFunction>) -> Self;
+    pub fn call<F>(&self, closure: F)
+        where F: FnOnce(&mut TaskContext, Handle<JsValue>, Handle<JsFunction>),
+              F: Send + 'static;
+}
+```
+
+The callback struct is backed by an `std::sync::Arc`. This allows the callback to be clone able and do be send across threads. The arc contains an opaque structure which is used to call the underlying C++ implementation.   
+The C++ implementation is thread safe and handles the asynchronous call to the main thread and the v8 stack setup.   
+Once the last clone of the `ThreadSafeCallback` is out of scope the `Arc` will drop the `ThreadSafeCallbackInner` which in turn will free any resources allocated by the C++ implementation.
+
+*Coming back to the example above:*   
+
+```rust
+let cb = ThreadSafeCallback::new(this, func);
+```
+`ThreadSafeCallback::new` allocates the C++ implementation, which stashes the provided `this`/`function` pair as well as the current `context` and initializes the async handle of `libuv` (`uv_async_init`). This has to occur on the main thread. After that the callback can be cloned and send across threads. 
+```rust
+// call into javascript
+cb.call(move |context, this, func| {
+    // call the js function
+    func.call(context, this, vec![cx.number(i).upcast()]);
+}
+```
+  
+When calling the `call` method, the given closure will be stashed in a thread safe queue and the main thread will be informed via `uv_async_send`. The main thread will call the function registered during the `uv_async_init` call.   
+After setting up the correct v8 scope every closure in the queue will be called with the current context and the stashed `this`/`function` pair.   
+The closure can now perform the Javascript call with the provided `this` and `func` values.
+Once `cb` goes out of scope the C++ implementation closes the async handle, which can only be done in the main thread, and delete itself.
+
+The `Arc` on the Rust side guarantees that the `close` method of the C++ implementation will be only called once and that no further `call`s to the C++ implementation are possible.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This will introduce some unsafe code and the implementation is tied directly to `libuv`
+as it uses the `uv_async_t` API.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+There are no real alternatives, calling into JavaScript from any (Rust) thread is a useful feature for bindings author.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Would it make sense to generalize the API e.g. take a `Vec<JsValue>` instead of a `this`/`function` pair?
```

Comment:
I don't think so. The only use case I can think of for this is to be able to stash some JS data for easy access when waking up. In practice, I think most of the time you would unwrap this data into native rust types before yielding.

If you must do this, the callback could capture the data and call back into rust, albeit with a small performance cost.